### PR TITLE
Run kubernetes integration tests inside of a pod and use kind to setup a kubernetes cluster

### PR DESCRIPTION
**Pull request by blakerouse**
## What does this PR do?

Refactors how `mage goIntegTest` run for metricbeat. Before each module was run inside of a docker container using `docker-compose`. Now depending on the module it will either run in docker or kubernetes. At the moment only the `kubernetes` module runs inside of kubernetes.

Each module is determined if it should run inside of docker or inside of kubernetes. If the module has a `docker-compose.yml` file then it will run in docker, if the module has a `kubernetes.yml` then it will run inside of kubernetes.

The docker runner stuff is the same so nothing really changes there, but the kubernetes runner is what is new. It works by executing the tests inside of a Pod running inside of the kubernetes.

If `KUBECONFIG` is defined in the environment then the runner just executes the tests against that defined kubernetes. If its not defined then kind will be used to setup a cluster and run the tests against that cluster.

Running the tests inside of kubernetes is handled by `KubeRemote` which uses the kubernetes client-go to setup all the requirements to get it running. Below breaks down the steps that are taken to get it running:

1. Generate a SSH private/public key.
2. Add public key as secret to kubernetes.
3. Add service account, cluster role, cluster role binding to give full access to all of kubernetes.
4. Deploy a Pod with an init container running `sshd-rsync` and the test executer container that starts once init container exits. The init container exits after the first SSH connection is made and disconnects.
5. While the init container is running port-forwarding is used to forward the SSH port to the local system.
6. `rsync` is ran locally to sync the beats repo to the init container sync directory which was mounted as an `emptyDir` and shared between the init container and exec container.
7. Init container exits and the execute container starts the test for the specific module in this case `kubernetes`.
8. Then the logs are tailed to the local terminal while the test runner is running.
9. Once the test is done all the added resources are cleaned up.

## Why is it important?

Currently all of the Kubernetes integration tests are skipped, this fixes them so they can actually be ran inside of a real Kubernetes cluster.

## Checklist

- [X] My code follows the style guidelines of this project
- [X] I have commented my code, particularly in hard-to-understand areas
- ~~I have made corresponding changes to the documentation~~
- ~~I have made corresponding change to the default configuration files~~
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [X] Ensure that Kubernetes integration tests are ran.

## How to test this PR locally

```
MODULE="kubernetes" mage goIntegTest
```

## Related issues

- Related elastic/beats#17538

### CODE REVIEW

**Comment by blakerouse** [comment]
@exekias @jsoriano @ChrsMark Whats your thoughts on running the intergration tests for Kubernetes using kind?

Feels like to me this works rather well and we could add more detailed integration tests this way. Kind also allows mounting of volumes, could possible use this with filebeat to access the log path and have filebeat connect to kubernetes API server to test autodiscovery.

At the moment this works as long as `kind` and `kubectl` is installed. There is still 1 test failing, but its not related to the actual running cluster, seems just like a test issue (still looking into it, thats why its in draft).

**Comment by blakerouse** [comment]
Been looking at the failed test it seems that `state_container` metricset is added a top level key of `container` to the event. It has both `kubernetes.container` and `container`, is that correct? I would think not unless it has to do with ECS, maybe?

```
(Pdb) sorted(self.de_dot(KUBERNETES_FIELDS).keys())
['@timestamp', 'agent', 'ecs', 'event', 'host', 'kubernetes', 'metricset', 'service']
(Pdb) sorted(evt.keys())
['@timestamp', 'agent', 'container', 'ecs', 'event', 'host', 'kubernetes', 'metricset', 'service']
```

**Comment by blakerouse** [comment]
Added more tests, got almost all metrcsets tested except for `event`, `volume`, `state_resourcequota`.

File: libbeat/tests/system/beat/kind.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+        Execute kubectl against the kind cluster.
+        """
+        if input is not None:
+            input = input.encode("utf-8")
+        kubecfg_path = cls.kind_kubecfg_path(build_path)
+        args = ["kubectl", "--kubeconfig", kubecfg_path] + args
+        return subprocess.run(args, capture_output=capture_output, check=check, input=input)
+
+    @classmethod
+    @contextlib.contextmanager
```
Comment (jsoriano) [inline libbeat/tests/system/beat/kind.py:100]: Nice, I didn't know it.

File: metricbeat/module/kubernetes/test_kubernetes.py [hunk 2]
```diff
@@ -4,55 +4,362 @@
...
+
+RESOURCE_QUOTA_MANIFEST = """\
+apiVersion: v1
+kind: ResourceQuota
+metadata:
+  name: object-counts
+spec:
+  hard:
+    configmaps: "99"
+"""
```
Comment (jsoriano) [inline metricbeat/module/kubernetes/test_kubernetes.py:179]: Move these manifests to files?

**Comment by exekias** [comment]
Thank you for starting this! I understand that this will work in any linux box with Docker installed? If that's the case this sounds like a nice approach to me!

any thoughts @jsoriano? In the past we discussed treating this similar to the cloud provider integration tests, but this looks simpler.

> Been looking at the failed test it seems that `state_container` metricset is added a top level key of `container` to the event. It has both `kubernetes.container` and `container`, is that correct? I would think not unless it has to do with ECS, maybe?
> 
> ```
> (Pdb) sorted(self.de_dot(KUBERNETES_FIELDS).keys())
> ['@timestamp', 'agent', 'ecs', 'event', 'host', 'kubernetes', 'metricset', 'service']
> (Pdb) sorted(evt.keys())
> ['@timestamp', 'agent', 'container', 'ecs', 'event', 'host', 'kubernetes', 'metricset', 'service']
> ```

This is probably related to some ECS mapping, where we put some info about the container under ECS.

**Comment by exekias** [comment]
@blakerouse do you know if this will allow us to access container logs?

File: metricbeat/module/kubernetes/test_kubernetes.py [hunk 3]
```diff
@@ -4,55 +4,362 @@
...
+        cls.kind_delete_cluster(cls.build_path)
+
+    @classmethod
+    def kube_state_metrics_hosts(cls):
+        """
+        kube-state-metrics host connection.
+        """
+        return ["localhost:30808"]
+
+    @unittest.skipUnless(KindMixin.kind_available() and metricbeat.INTEGRATION_TESTS, "integration test")
```
Comment (jsoriano) [inline metricbeat/module/kubernetes/test_kubernetes.py:219]: We can also use `TEST_TAGS` (see https://github.com/elastic/beats/pull/17075) to select when to run these tests. This way we avoid skipping the tests by mistake because `kind` is not properly installed in some CI worker.
```suggestion
    @unittest.skipUnless(metricbeat.INTEGRATION_TESTS, "integration test")
    @metricbeat.tag('kubernetes')
```
These tests will need to be run then with `TEST_TAGS=kubernetes`

File: metricbeat/Makefile [hunk 4]
```diff
@@ -83,3 +86,10 @@ integration-tests: ## @testing Run golang integration tests.
 integration-tests: prepare-tests mage
 	rm -f docker-compose.yml.lock
 	mage goIntegTest
+
+.PHONY: prepare-tests
+prepare-tests:
+	$(MAKE) -f ${ES_BEATS}/libbeat/scripts/Makefile prepare-tests
+	# kind and kubectl is needed to run the kubernetes system-tests.
+	go ${INSTALL_CMD} ${KIND_TOOL_REPO}
+	# TODO: Either install kubectl or convert to python kubernetes client.
```
Comment (jsoriano) [inline metricbeat/Makefile:95]: At some point this will need to be migrated to mage.

**Comment by jsoriano** [comment]
Btw, it may be also good to add some environment variable that can be used to run the tests using an existing cluster. For example think on making that if `KUBECONFIG` environment variable is set, the test doesn't try to start a cluster, and instead it uses the config in `KUBECONFIG`.
This would allow us to run the same tests against different Kubernetes providers.

**Comment by blakerouse** [comment]
Thanks for the feedback! I am still working on this because currently the testing is all over the place.

Travis tests with minikube (by deploying the manifests, without even checking if they actually deployed), Jenkins tests with kind (by deploying the manifests, without even checking if they actually deployed), and the Kubernetes python integration tests that where disabled and didn't work.

My goal is to get it all working with Kind and across multiple versions of Kubernetes.

@jsoriano Will update to use `TEST_TAGS`, I was wanting a way to selectively run these tests and this is perfect! As for the `KUBECONFIG` idea the current issue would be that the tests would most likely fail because it needs the cluster on the local system to actually pass the tests. It also doesn't currently build docker images to run inside of Kind, it runs metricbeat outside of the cluster on the host. More work will need to be done to build the images and run them inside of the cluster. (I will see how hard this would be, because it would allow us to do what your suggesting, which would allow use to test inside of other cloud providers easily.

@exekias I believe it will allow use to get access to the logs because you can bind mount into the k8s node with kind allowing the host to get the logs from the containers. If we transition to building the images and deploying those into the cluster, then it definitely could always access them.

**Comment by exekias** [comment]
I would avoid building + pushing the image if we don't really need to, that's something we will test in e2e tests anyway (ie helm chart tests are doing this already).

**Comment by blakerouse** [comment]
If we are going to avoid that and save those tests for e2e, then I will not add the `KUBECONFIG` check as the tests will require the usage of kind.

I also think we should remove the `test` target in `deploy/kubernetes/Makefile` and add the test inside of the kubernetes metricbeat integration tests. That way its all in once place and not divided up. I think the test really should just verify the schema of the manifest, because the images will not be built.

@exekias wydt?

**Comment by exekias** [comment]
> If we are going to avoid that and save those tests for e2e, then I will not add the `KUBECONFIG` check as the tests will require the usage of kind.

I understand the use of KUBECONFIG aims to be able to launch tests against a running you may have (when developing). KUBECONFIG should work also when running from outside the cluster, isn't it?

> I also think we should remove the `test` target in `deploy/kubernetes/Makefile` and add the test inside of the kubernetes metricbeat integration tests. That way its all in once place and not divided up. I think the test really should just verify the schema of the manifest, because the images will not be built.

I would hope we can keep tests separted by concern, for instance, Filebeat & Metricbeat tests are different, or Kubernetes module vs autodiscover. They belong in different places in the code. As we move towards the agent & integrations living somewhere else, I would expect this split to happen anyway at some point, so it's probably better to keep things separated today.

That said, I don't see anything wrong with sharing the code to launch kind across the board, for instance

**Comment by blakerouse** [comment]
> I understand the use of KUBECONFIG aims to be able to launch tests against a running you may have (when developing). KUBECONFIG should work also when running from outside the cluster, isn't it?

To expose `kube-state-metrics` with kind the code uses a service with a nodeport and `kind create cluster` that maps that port to the local system. Without this specifics the `kube-state-metrics` tests will fail. I could make it work with `KUBECONFIG` and skip the `kube-state-metrics` if using that mode.

**Comment by blakerouse** [comment]
@exekias @jsoriano I have actually updated the PR to run the tests with `goIntegTest` as it seemed the future was to move all tests to be ran with go instead of python. This meant changing the integration test runner quite a bit to allow this, but it enables us to add lots of kubernetes tests without having to worry about what was used to create the cluster (the runner will use Kind, if no cluster is defined).

This means we can easily run integration tests against GKE, AKS, EKS, or others without having to adjust the integration tests, as the tests can assume they are running inside of a kubernetes pod that has all rights to the cluster itself.

**Comment by blakerouse** [comment]
I also forgot to mention that this currently does not test all metricsets in the kubernetes module. The branch has gotten so big that I wanted to stop adding and just get it working and merged before adding more, as the branch is already huge!

But after this is merged I will be adding more tests to cover all metricsets in kubernetes, add in field coverage, add tests for filebeat and heartbeat, add tests for autodiscover for metricbeat, filebeat, and heartbeat.

File: dev-tools/make/mage.mk [hunk 5]
```diff
@@ -1,13 +1,53 @@
...
-
-.PHONY: mage
-mage:
-ifndef MAGE_PRESENT
-	@echo Installing mage $(MAGE_VERSION) from vendor dir.
-	@go install -mod=vendor -ldflags="-X $(MAGE_IMPORT_PATH)/mage.gitTag=$(MAGE_VERSION)" ${MAGE_IMPORT_PATH}/...
-	@-mage -clean
-endif
-	@true
+# This is a minimal Makefile for Beats that are built with Mage. Its only
```
Comment (exekias) [inline dev-tools/make/mage.mk:1]: it looks like this could go in a different PR?

File: dev-tools/mage/integtest.go [hunk 6]
```diff
@@ -242,19 +396,77 @@ func IsInIntegTestEnv() bool {
...
+	types := map[string]string{}
+	dockerFile := filepath.Join(cwd, dir, "docker-compose.yml")
+	if _, err := os.Stat(dockerFile); !os.IsNotExist(err) {
+		types[TestingEnvDocker] = dockerFile
+	}
+	kubeFile := filepath.Join(cwd, dir, "kubernetes.yml")
+	if _, err := os.Stat(kubeFile); !os.IsNotExist(err) {
+		types[TestingEnvKubernetes] = kubeFile
+	}
+	return &IntegrationEnv{types}
```
Comment (exekias) [inline dev-tools/mage/integtest.go:419]: @jsoriano do you see a good fit for using terraform here too? :innocent:

Comment (jsoriano) [reply dev-tools/mage/integtest.go:419]: Yes, this will be definitely an option if we move terraform scenarios to modules :slightly_smiling_face:

**Comment by exekias** [comment]
Thank you for digging deeper, this is looking really promising! Just a few comments:

- I understand that by now this would only work when you use `MODULE=kubernetes`, right? to be clear, this would be good as we can handle that from the CI perspective :)
- It would be nice to split the concerns of this PR, for instance moving the refactor on Makefile and mage-install to a new one, I think @andrewkroh has some plans about that

**Comment by blakerouse** [comment]
@exekias I am currently working on splitting the mage work out into its own PR.

It does not require `MODULE=kubernetes` it will run the kubernetes tests when running `mage goIntegTest` as long as either `KUBECONFIG` or `kind` is present.

File: dev-tools/mage/integtest.go (hunk 6 above)
Comment (blakerouse) [reply dev-tools/mage/integtest.go:419]: Yes, I think this would be a great place to connect with terraform, as I would bet that most `*.tf` will be module specific.

Could even been combined together:

1. Deploy GKE with Terraform
2. Run kubernetes integration tests in GKE.

That was the exact reason I put a lot of work in running the tests inside of a Pod inside of Kubernetes to allow integration tests with cloud kubernetes deployments.

**Comment by blakerouse** [comment]
#17799 splits out the mage work, once that merges I will rebase and cleanup this PR.

**Comment by jsoriano** [comment]
> #17799 splits out the mage work, once that merges I will rebase and cleanup this PR.

Thanks! I will take a look.

File: dev-tools/mage/integtest.go [hunk 7]
```diff
@@ -186,53 +218,175 @@ func runInIntegTestEnv(mageTarget string, test func() error, passThroughEnvVars
...
+		if err != nil {
+			return err
+		}
+		// Uses `os.Stdout` directly as its output should always be shown.
+		err = remote.Run(remoteEnv, os.Stdout, stdErr, magePath, mageTarget)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
```
Comment (jsoriano) [inline dev-tools/mage/integtest.go:389]: Split this in different methods please, at least one for docker compose and another one for kubernetes.

File: dev-tools/mage/integtest.go [hunk 8]
```diff
@@ -242,19 +396,77 @@ func IsInIntegTestEnv() bool {
...
+}
+
+// GetTypeSource returns the source file that determined that type.
+func (i *IntegrationEnv) GetTypeSource(t string) (string, bool) {
+	source, ok := i.types[t]
+	return source, ok
+}
+
+func haveIntegTestEnvRequirements(testEnv *IntegrationEnv) error {
+	if testEnv.HasType(TestingEnvDocker) || testEnv.HasType(TestingEnvKubernetes) {
```
Comment (jsoriano) [inline dev-tools/mage/integtest.go:435]: I see many `if testEnv.HasType(...)`, this will make problematic to add another provider (what will surely happen with terraform, or with future cloud-based kubernetes providers).
We could make `testEnv` to contain a list of "types" or providers, and each type has its way of check if it should be skipped, check their requirements, to execute and to cleanup. This way we can iterate on them and execute each operation instead of adding checks for each type.

File: dev-tools/mage/integtest.go [hunk 9]
```diff
@@ -319,6 +539,30 @@ func dockerComposeProjectName() string {
...
+	version = strings.NewReplacer(".", "_").Replace(version)
+
+	clusterName := "{{.BeatName}}_{{.Version}}_{{.ShortCommit}}-{{.StackEnvironment}}"
+	clusterName = MustExpand(clusterName, map[string]interface{}{
+		"StackEnvironment": StackEnvironment,
+		"ShortCommit":      commit[:10],
+		"Version":          version,
+	})
+	return clusterName
+}
```
Comment (jsoriano) [inline dev-tools/mage/integtest.go:564]: Move this to some kubernetes specific file?

File: dev-tools/mage/integtest.go (hunk 6 above)
Comment (ChrsMark) [reply dev-tools/mage/integtest.go:419]: Just for reference and since maybe we would like to test on Openshift too in the future: Openshift installation on GCP is achieved with their installer which uses `terafform` underneath (see https://github.com/elastic/beats/issues/17516#issuecomment-617038764).

File: metricbeat/module/kubernetes/apiserver/_meta/fields.yml [hunk 10]
```diff
@@ -147,7 +147,7 @@
         - name: duration.us.count
           type: long
           description: Request duration, number of operations
-        - name: duration.us.bucket.*
+        - name: duration.bucket.us.*
```
Comment (ChrsMark) [inline metricbeat/module/kubernetes/apiserver/_meta/fields.yml:150]: `us` goes first in order to have the buckets in the same "object" with `duration.us.sum` and `duration.us.count`. Is this change intended?

Comment (blakerouse) [reply metricbeat/module/kubernetes/apiserver/_meta/fields.yml:150]: Yes because without this change the test I added for kubernetes was failing. I will look into why its failing, maybe the code needs to be fixed not the field.

Comment (blakerouse) [reply metricbeat/module/kubernetes/apiserver/_meta/fields.yml]: I was able to remove this for now, as I removed the field validation from the tests. I want to add that in a follow up PR, but this PR is already way to big.

File: dev-tools/mage/integtest.go (hunk 7 above)
Comment (blakerouse) [reply dev-tools/mage/integtest.go]: I have completely separated this code and converted everything into different pieces.

See `DockerIntegrationTester` and `KubernetesIntegrationTester` in latest commit.

File: dev-tools/mage/integtest.go (hunk 8 above)
Comment (blakerouse) [reply dev-tools/mage/integtest.go]: See latest commit for the solution to this problem.

A new runner can be easily registered with `RegisterIntegrationTester`.

I would like to add a scenarios runner that extends on this even more, that would allow things like inside of the kubernetes module:

```
module/kubernetes/integration_scenarios/aks/terraform.tf
module/kubernetes/integration_scenarios/aks/kubernetes.yml
module/kubernetes/integration_scenarios/eks/terraform.tf
module/kubernetes/integration_scenarios/eks/kubernetes.yml
module/kubernetes/integration_scenarios/gke/terraform.tf
module/kubernetes/integration_scenarios/gke/kubernetes.yml
```

Then the scenario runner would run the tests in each scenario `aks`, `eks`, `gke`. First using a `TerraformTestSetupStep` followed by the `KubernetesIntegrationTester`.

Execution order would be:

```
ScenarioIntegrationTester.Test
  - AWS
     - Terraform Step
     - Kubernetes Tester
  - EKS
     - Terraform Step
     - Kubernetes Tester
  - GKE
     - Terraform Step
     - Kubernetes Tester
```

Everything in the PR is designed to head down that path, looking forward to working on future branches to enable that work.

File: dev-tools/mage/integtest.go (hunk 9 above)
Comment (blakerouse) [reply dev-tools/mage/integtest.go:326]: See `integtest_kubernetes.go`

**Comment by blakerouse** [comment]
@jsoriano Okay this is completely ready for a full review.

**Comment by blakerouse** [comment]
Seems like some of those test failures are miss reported from Jenkins, I only see one random test failure from filebeat and the generators test (which I am fixing in a seperate branch).

The kubernetes tests ran and passed! Finally!

File: dev-tools/mage/kuberemote.go [hunk 11]
```diff
@@ -0,0 +1,612 @@
...
+	if err != nil {
+		return err
+	}
+
+	rsh := fmt.Sprintf("/usr/bin/ssh -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o LogLevel=ERROR -p %d -i %s", port, privateKeyFile)
+	args := []string{
+		"--rsh", rsh,
+		"-a", fmt.Sprintf("%s/", r.syncDir),
+		fmt.Sprintf("root@localhost:%s", r.destDir),
+	}
```
Comment (jsoriano) [inline dev-tools/mage/kuberemote.go:259]: On local development copying the whole working copy can be too much, but it seems to work fine by now, we can evaluate alternatives later if this becomes a problem.
Would it be possible to mount the directory in the kind container, and from it into the pod?

File: dev-tools/mage/integtest_kind.go [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+// to be used. It cannot be autodiscovered for usage.
+func (m *KindIntegrationTestStep) Use(dir string) (bool, error) {
+	return false, nil
+}
+
+// Setup ensures that a kubernetes cluster is up and running.
+//
+// If `KUBECONFIG` is already deinfed in the env then it will do nothing.
+func (m *KindIntegrationTestStep) Setup(env map[string]string) error {
+	_, exists := env["KUBECONFIG"]
```
Comment (jsoriano) [inline dev-tools/mage/integtest_kind.go:51]: It could be nice to add a README into the kubernetes module with brief instructions about the use of `KUBECONFIG` for local testing. Maybe with instructions for manually starting a local cluster with kind or minikube.

File: dev-tools/mage/kuberemote.go [hunk 13]
```diff
@@ -0,0 +1,612 @@
...
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/tools/portforward"
+	watchtools "k8s.io/client-go/tools/watch"
+	"k8s.io/client-go/transport/spdy"
```
Comment (jsoriano) [inline dev-tools/mage/kuberemote.go:52]: Out of curiosity, would it be possible to use the transports in `k8s.io/client-go/transport` or `k8s.io/client-go/rest`? I guess that these ones use HTTP/2, so it shouldn't be so different in performance to SPDY. They are also probably more maintained, and we wouldn't need to add the SPDY dependencies.

File: dev-tools/mage/common.go [hunk 14]
```diff
@@ -188,7 +188,7 @@ func (info *DockerInfo) IsBoot2Docker() bool {
 // HaveDocker returns an error if docker is unavailable.
 func HaveDocker() error {
 	if _, err := GetDockerInfo(); err != nil {
-		return errors.Wrap(err, "docker is not available")
+		return fmt.Errorf("docker is not available")
```
[the remark this answers is not in the document]
Comment (blakerouse) [reply dev-tools/mage/common.go:191]: I really don't have a reason for why I made that change, seems I was doing something different there then changed it back wrongly.

Reverted the change to wrap the error.

File: Jenkinsfile [hunk 15]
```diff
@@ -801,14 +801,14 @@ def dumpFilteredEnvironment(){
...
         withGithubNotify(context: "K8s ${v}") {
           withBeatsEnv(false) {
-            sh(label: "Install k8s", script: """
-              eval "\$(gvm use ${GO_VERSION} --format=bash)"
-              .ci/scripts/kind-setup.sh
-            """)
-            sh(label: "Kubernetes Kind",script: "make KUBECONFIG=\"\$(kind get kubeconfig-path)\" -C deploy/kubernetes test")
+            sh(label: "Install kind", script: ".ci/scripts/install-kind.sh")
+            sh(label: "Install kubectl", script: ".ci/scripts/install-kubectl.sh")
+            sh(label: "Integration tests", script: "MODULE=kubernetes make -C metricbeat integration-tests")
```
[the remark this answers is not in the document]
Comment (blakerouse) [reply Jenkinsfile:809]: Used make because the PWD is the root of beats, so using make allows the `make -C metricbeat` to easily go into that directory.

File: dev-tools/mage/gotest.go [hunk 16]
```diff
@@ -137,33 +138,43 @@ func DefaultTestBinaryArgs() TestBinaryArgs {
...
+			continue
 		}
+		if module != "" && module != fi.Name() {
+			continue
+		}
+		foundModule = true
 
-		modulesFileInfo, err := ioutil.ReadDir("./module")
+		// Set MODULE because only want that modules tests to run inside the testing environment.
+		os.Setenv("MODULE", fi.Name())
```
[the remark this answers is not in the document]
Comment (blakerouse) [reply dev-tools/mage/gotest.go:158]: It's important to set MODULE here because at the top level we want to execute docker or kubernetes unique per module. If we didn't add this here then every module test would run inside of the testing environment which we do not want. The test runner uses the MODULE from the running environment to pass into the underlying test environment using `-e` in the docker case and `env: {}` in the kubernetes case.

File: dev-tools/mage/gotest.go [hunk 17]
```diff
@@ -137,33 +138,43 @@ func DefaultTestBinaryArgs() TestBinaryArgs {
...
+		err = runners.Test("goIntegTest", func() error {
 			err := GoTest(ctx, GoTestIntegrationArgsForModule(fi.Name()))
 			if err != nil {
-				failed = true
+				return errors.Wrapf(err, "integration tests failed for module %s", fi.Name())
 			}
+			return nil
+		})
+		if err != nil {
+			return err
```
[the remark this answers is not in the document]
Comment (blakerouse) [reply dev-tools/mage/gotest.go:171]: Fixed.

File: dev-tools/mage/integtest.go [hunk 18]
```diff
@@ -19,334 +19,331 @@ package mage
...
-	"github.com/magefile/mage/sh"
+	"github.com/pkg/errors"
 )
 
 const (
 	// BEATS_DOCKER_INTEGRATION_TEST_ENV is used to indicate that we are inside
-	// of the Docker integration test environment (e.g. in a container).
-	beatsDockerIntegrationTestEnvVar = "BEATS_DOCKER_INTEGRATION_TEST_ENV"
+	// of the integration test environment.
+	insideIntegrationTestEnvVar = "BEATS_DOCKER_INTEGRATION_TEST_ENV"
```
[the remark this answers is not in the document]
Comment (blakerouse) [reply dev-tools/mage/integtest.go:34]: Fixed.

File: Jenkinsfile (hunk 15 above)
[the remark this answers is not in the document]
Comment (andrewkroh) [reply Jenkinsfile:809]: `mage -d` works like `make -C`.

File: metricbeat/module/kubernetes/apiserver/apiserver_integration_test.go [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+	"github.com/stretchr/testify/assert"
+
+	mbtest "github.com/elastic/beats/v7/metricbeat/mb/testing"
+	"github.com/elastic/beats/v7/metricbeat/module/kubernetes/test"
+)
+
+func TestFetchMetricset(t *testing.T) {
+	config := test.GetAPIServerConfig(t, "apiserver")
+	metricSet := mbtest.NewReportingMetricSetV2Error(t, config)
+	events, errs := mbtest.ReportingFetchV2Error(metricSet)
```
[the remark this answers is not in the document]
Comment (blakerouse) [reply metricbeat/module/kubernetes/apiserver/apiserver_integration_test.go:34]: Updated.

**Comment by andrewkroh** [comment]
Does this work with the Kubernetes built into the Docker for Mac install? I think a lot of us develop on Mac with this setup so if that just worked it would be nice.

File: dev-tools/mage/kuberemote.go [hunk 20]
```diff
@@ -0,0 +1,612 @@
...
+			rbacv1.PolicyRule{
+				Verbs:     []string{"*"},
+				APIGroups: []string{"*"},
+				Resources: []string{"*"},
+			},
+			rbacv1.PolicyRule{
+				Verbs:           []string{"*"},
+				NonResourceURLs: []string{"*"},
+			},
+		},
```
[the remark this answers is not in the document]
Comment (blakerouse) [reply dev-tools/mage/kuberemote.go:375]: Yeah I like that idea, maybe another time.

File: dev-tools/mage/kuberemote.go (hunk 11 above)
Comment (blakerouse) [reply dev-tools/mage/kuberemote.go:259]: Possibly but the benefit of this is greater, because it allows the cluster that is going to run the tests to not even be local to the test runner. So the integration tests could be started from your local computer but be running on GKE.

File: dev-tools/mage/integtest_kind.go (hunk 12 above)
Comment (blakerouse) [reply dev-tools/mage/integtest_kind.go:51]: Added a readme explaining how to run the integration tests for kubernetes.

**Comment by blakerouse** [comment]
@andrewkroh Yes as long as you have `docker`, `kind` and `kubectl` installed on your mac it jsut works. I developed this and tested it on a mac.

File: dev-tools/mage/kuberemote.go (hunk 13 above)
Comment (blakerouse) [reply dev-tools/mage/kuberemote.go:52]: SPDY is how kubectl does it for port-forwarding so that is how I implemented it.

https://github.com/kubernetes/kubectl/blob/master/pkg/cmd/portforward/portforward.go#L138

File: dev-tools/mage/kuberemote.go [hunk 21]
```diff
@@ -0,0 +1,612 @@
...
+}
+
+// rsync performs the rsync of sync directory to destination directory inside of the pod.
+func (r *KubeRemote) rsync(port uint16, stdout, stderr io.Writer) error {
+	privateKeyFile, err := createTempFile(r.privateKey)
+	if err != nil {
+		return err
+	}
+
+	rsh := fmt.Sprintf("/usr/bin/ssh -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o LogLevel=ERROR -p %d -i %s", port, privateKeyFile)
```
Comment (andrewkroh) [inline dev-tools/mage/kuberemote.go:254]: Does the path to `ssh` need to be hardcoded? Any chance this will work for Windows developers?

File: dev-tools/mage/gotest.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (blakerouse) [reply dev-tools/mage/gotest.go]: Yes I have updated it so its passed in instead of set in the current process.

File: dev-tools/mage/integtest_kubernetes.go [hunk 22]
```diff
@@ -0,0 +1,164 @@
...
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"github.com/magefile/mage/mg"
+)
+
+func init() {
+	RegisterIntegrationTester(&KubernetesIntegrationTester{})
```
[the remark this answers is not in the document]
Comment (blakerouse) [reply dev-tools/mage/integtest_kubernetes.go]: Done!

File: dev-tools/mage/kuberemote.go (hunk 21 above)
Comment (blakerouse) [reply dev-tools/mage/kuberemote.go:254]: It does not, I removed the hardcode path. If on windows they had `rsync` and `ssh` in PATH then it most likely would work.

**Comment by blakerouse** [comment]
Okay all comments have either been answered or resolved. Anyway I can get a +1 to get this landed. Blocking me adding even more kubernetes tests.

**Comment by jsoriano** [comment]
I have seen after updating https://github.com/elastic/beats/pull/17480 with master that with this change integration tests for modules without docker compose or kubernetes files are not executed, not so serious because it only affects some system tests, and cloud tests, that are not merged yet.

**Comment by blakerouse** [comment]
@jsoriano Could you file a bug about what is not working? Be happy to fix it.

**Comment by jsoriano** [comment]
> @jsoriano Could you file a bug about what is not working? Be happy to fix it.

@blakerouse I opened a PR with a temptative fix https://github.com/elastic/beats/pull/18125, let me know what you think.

**Comment by andrewkroh** [comment]
@blakerouse The Metricbeat generator test is failing in master. Do you think it's related to your changes?

https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats-beats-mbp/detail/master/1571/pipeline/573

```
[2020-04-30T16:41:41.466Z] beatpath/testmetricbeat imports

[2020-04-30T16:41:41.466Z] 	github.com/elastic/beats/v7/dev-tools/mage imports

[2020-04-30T16:41:41.466Z] 	k8s.io/client-go/kubernetes imports

[2020-04-30T16:41:41.466Z] 	k8s.io/client-go/discovery imports

[2020-04-30T16:41:41.466Z] 	github.com/googleapis/gnostic/OpenAPIv2: module github.com/googleapis/gnostic@latest found (v0.4.1), but does not contain package github.com/googleapis/gnostic/OpenAPIv2
```

**Comment by blakerouse** [comment]
@andrewkroh I am working on this here - https://github.com/elastic/beats/pull/18086

The latest run I got with that MP showed something different.
